### PR TITLE
fix: add JS to manipulate the content for generic styling to work

### PR DESCRIPTION
**Pull request by philschatz**
This addresses openstax/unified#285.

Code taken from [webview/src/scripts/modules/media/body/body.coffee#L123](https://github.com/openstax/webview/blob/f95b1d0696a70f0b61d83a85c173102e248354cd/src/scripts/modules/media/body/body.coffee#L123)

# Acceptance Tests

derive these from "Acceptance Criteria" during "Design" phase

**title**: Classed Notes styled properly
**categories**: Book Content
> GIVEN User is on page 6.5 Periodic Variations in Element Properties {base_url}/books/business-ethics/pages/1-1-being-a-professional-of-integrity#fs-idm400154912
> WHEN the page is loaded and scrolled to feature Link to Learning
> THEN feature has the appropriate padding above the title
> AND feature title is in all caps
> AND feature displays horizontal line below the title as shown in <https://cnx.org/contents/kUrGbuHs@4.1:FsaaJM1v@3/1-1-Being-a-Professional-of-Integrity#fs-idm400154912>

image

**title**: Lists not bulleted
**categories**: Book Content
> GIVEN User is on Preface {base_url}/books/chemistry-2e/pages/preface#eip-951
> WHEN the page is loaded and scrolled to text Pedagogical foundation and features
> THEN list displays properly (no bullets), as shown in <https://cnx.org/contents/f8zJz5tx@6.2:E4Y07W7Q@6/Preface#eip-951>

image

**title**: Answer styled correctly
**categories**: Book Content
> GIVEN User is on page 6.1 Electromagnetic Energy {base_url}/books/chemistry-2e/pages/6-1-electromagnetic-energy#fs-idp125103008
> WHEN the page is loaded and scrolled to Answer
> THEN text ANSWER is in all caps
> AND the appropriate padding is present 
> AND separator line is displayed under the text as shown in <https://cnx.org/contents/f8zJz5tx@6.2:QepiZWK9@5/6-1-Electromagnetic-Energy#fs-idp125103008>

image

**title**: Add space between paragraphs in Example
**categories**: Book Content
> GIVEN User is on page 6.4 Electronic Structure of Atoms (Electron Configurations) {base_url}/books/chemistry-2e/pages/6-4-electronic-structure-of-atoms-electron-configurations#fs-idp31411280
> WHEN the page is loaded and scrolled to feature Example 6.11
> THEN there is a space between paragraphs as shown in <https://cnx.org/contents/f8zJz5tx@6.2:Yn5aW5Sp@10/6-4-Electronic-Structure-of-Atoms-Electron-Configurations#fs-idp31411280>

image

### CODE REVIEW

File: src/app/components/MainContent.tsx [hunk 1]
```diff
@@ -1,17 +1,118 @@
...
+    rootEl.querySelectorAll('figure > figcaption').forEach((el) => {
+      const parent = assertDefined(el.parentElement, 'figcaption parent should always be defined');
+      parent.classList.add('ui-has-child-figcaption');
+      parent.appendChild(el)
+    });
+  }
+
+  // Add nofollow to external user-generated links
+  public addNoFollow(rootEl: Element) {
+    rootEl.querySelectorAll('a[href^="http:"], a[href^="https:"], a[href^="//"]')
```
Comment (TomWoodward) [inline src/app/components/MainContent.tsx]: i'm thinking maybe this is unnecessarily specific, not sure there is a reason to allow following on any links in the content

File: src/test/fixtures/contents/testbook1-shortid:testpage1-shortid [hunk 2]
```diff
@@ -3,5 +3,5 @@
...
...n lacus. Nullam pretium mollis laoreet. Ut interdum dui sit amet elementum dignissim. Etiam in nibh vitae ligula fringilla aliquet quis vel ex. Nullam quis elit libero. In aliquet laoreet arcu at posuere. Duis consectetur et dolor non accumsan. Fusce enim est, dignissim sed mauris at, ornare cursus ligula. Nullam tincidunt eu purus vel lacinia. Praesent varius, arcu sed facilisis elementum, orci metus vestibulum nisl, sed sodales urna eros ullamcorper turpis. Donec mollis ac massa ac elementum. Nulla maximus purus sed orci efficitur, ut porta mauris elementum.</p><p id=\"p9\">Vestibulum blandit enim quis sapien ultrices pretium. Cras dapibus ornare fringilla. Duis ut laoreet dolor. Proin ullamcorper molestie dui eu vestibulum. Nullam convallis mollis mauris nec accumsan. Proin egestas enim non nisl suscipit, ut feugiat felis pharetra. In faucibus fringilla ipsum vel sollicitudin. Maecenas blandit augue odio, ac mattis justo feugiat sit amet. Etiam nec molestie eros. Sed gravida velit libero, at pretium diam venenatis sit amet. Donec tellus leo, faucibus eu sapien nec, mollis efficitur lorem. Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p></section></div></body></html>"
```
Comment (TomWoodward) [inline src/test/fixtures/contents/testbook1-shortid:testpage1-shortid]: why is this chagned?

**Review by TomWoodward — CHANGES_REQUESTED** [review]
it would be better to add all this logic to content/components/Page since app/components/MainContent is outside the content module, the purpose of the main content is for the "skip to content" link

File: src/app/components/MainContent.tsx [hunk 3]
```diff
@@ -1,17 +1,118 @@
...
+    .forEach((el) => el.setAttribute('rel', 'nofollow'));
```
[the remark this answers is not in the document]
Comment (philschatz) [reply src/app/components/MainContent.tsx]: Agreed

File: src/test/fixtures/contents/testbook1-shortid:testpage1-shortid (hunk 2 above)
Comment (philschatz) [reply src/test/fixtures/contents/testbook1-shortid:testpage1-shortid]: I changed this fixture in order to exercise the code in the PR. I could also create a separate fixture file.

Comment (TomWoodward) [reply src/test/fixtures/contents/testbook1-shortid:testpage1-shortid]: in what way did you modify it? this is the page that the screenshot spec uses to exercise sims and math, as long as that stuff is still there its probably fine

Comment (philschatz) [reply src/test/fixtures/contents/testbook1-shortid:testpage1-shortid]: I opted to write individual tests instead. Mucking with fixture files used by multiple tests seemed to be a bad approach.

**Comment by TomWoodward** [comment]
> The PR diff size of 5148 lines exceeds the maximum allowed for the inline comments feature.

image

wat

File: src/app/content/components/Page.tsx [hunk 4]
```diff
@@ -31,7 +32,7 @@ interface PropTypes {
 }
 
 export class PageComponent extends Component<PropTypes> {
-  public container: Element | undefined | null;
+  public container: Element | undefined | null = null;
```
Comment (TomWoodward) [inline src/app/content/components/Page.tsx]: it'll default to undefined though, why change this?

File: src/app/content/components/Page.tsx [hunk 5]
```diff
@@ -98,6 +101,84 @@ export class PageComponent extends Component<PropTypes> {
...
+  private fixLists(rootEl: Element) {
+    // Copy data-mark-prefix and -suffix from ol to li so they can be used in css
+    rootEl.querySelectorAll(`ol[data-mark-prefix] > li, ol[data-mark-suffix] > li,
+    [data-type="list"][data-list-type="enumerated"][data-mark-prefix] > [data-type="item"],
+    [data-type="list"][data-list-type="enumerated"][data-mark-suffix] > [data-type="item"]`).forEach((el) => {
+      const parent = assertDefined(el.parentElement, 'list parent should always be defined');
+      const markPrefix = parent.getAttribute('data-mark-prefix');
+      const markSuffix = parent.getAttribute('data-mark-suffix');
+      if (markPrefix) { el.setAttribute('data-mark-prefix', markPrefix); }
+      if (markSuffix) { el.setAttribute('data-mark-suffix', markSuffix); }
```
Comment (TomWoodward) [inline src/app/content/components/Page.tsx]: are these branches maybe where your coverage is failing?

**Review by TomWoodward — APPROVED** [review]
this looks fine, later today i'll try to help track down the coverage issues

**Comment by TomWoodward** [comment]
@philschatz i think i fixed the code coverage, there is an html report that has more details than the cli output

one remaining comment about the types https://github.com/openstax/rex-web/pull/140/files/5c78f6434917627bed8faf0a220d39c47e97c232#diff-c103535bac1a28425b79861e889a5275 but otherwise good to go

**Comment by philschatz** [comment]
oops, I forgot to push my commit. The change was not necessary, I just don't like `undefined` being an expected value.

**Comment by TomWoodward** [comment]
if i had read your change more carefully i wouldn't have commented. i thought you had changed 

```diff
- public container: Element | undefined;
+ public container: Element | undefined | null = null;
```

but now i see that the null type was already in there. i agree that that is bad.

its actually doing it wrong anyway the new good way is to use `React.createRef`

lets just move on with our lives for now

**Pull request merged**